homeassistant/components/emulated_kasa/__init__.py

```diff
@@ -0,0 +1,104 @@
+"""Support for local power state reporting of entities by emulating TP-Link Kasa smart plugs."""
+import logging
+from time import time
+
+from sense_energy import PlugInstance, SenseLink
+import voluptuous as vol
+
+import homeassistant.components as comps
+from homeassistant.const import (
+    ATTR_FRIENDLY_NAME,
+    CONF_ENTITIES,
+    CONF_NAME,
+    EVENT_HOMEASSISTANT_STOP,
+)
+from homeassistant.core import HomeAssistant
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.template import Template
+
+from .const import CONF_POWER, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_ENTITY_SCHEMA = vol.Schema(
+    {
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Required(CONF_POWER): vol.Any(
+            vol.Coerce(float),
+            cv.template,
+        ),
+    }
+)
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ENTITIES): vol.Schema(
+                    {cv.entity_id: CONFIG_ENTITY_SCHEMA}
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Sense component."""
+    hass.data.setdefault(DOMAIN, {})
+    conf = config.get(DOMAIN)
+    if not conf:
+        return True
+    hass.data[DOMAIN][CONF_ENTITIES] = conf.get(CONF_ENTITIES, {})
+
+    def devices():
+        """Drvices to be emulated."""
+        yield from get_plug_devices(hass)
+
+    server = SenseLink(devices)
+
+    async def stop_emulated_kasa(event):
+        await server.stop()
+
+    try:
+        await server.start()
+    except OSError as error:
+        _LOGGER.error("Failed to create UDP server at port 9999: %s", error)
+    else:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, stop_emulated_kasa)
+
+    return True
+
+
+def get_plug_devices(hass):
+    """Produce list of plug devices from config entities."""
+    entity_ids = hass.data[DOMAIN][CONF_ENTITIES]
+    for entity_id in entity_ids:
```

Comment:
Actually it doesn't look like we need to store `conf.get(CONF_ENTITIES, {})` in `hass.data[DOMAIN][CONF_ENTITIES]` since nothing else it using it.  Can we just pass it in here as `entity_config`?

```suggestion
    for entity_id, entity_config in entity_config:
```